client/state/data-layer/wpcom/sites/jitm/index.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * Internal dependencies
+ */
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { isJetpackSite } from 'state/sites/selectors';
+import { getCurrentUserId } from 'state/current-user/selectors';
+import { SECTION_SET, SELECTED_SITE_SET, JITM_SET } from 'state/action-types';
+import config from 'config';
+
+/**
+ * Poor man's process manager
+ * @type {{
+ * hasInitializedSites: boolean,
+ * hasInitializedSection: boolean,
+ * lastSection: null|string,
+ * lastSite: null|string,
+ * }}
+ */
+const process = {
+	hasInitializedSites: false,
+	hasInitializedSection: false,
+	lastSection: null,
+	lastSite: null,
+};
+
+/**
+ * Processes the current state and determines if it should fire a jitm request
+ * @param {object} state The current state
+ * @param {function} dispatch The redux dispatch function
+ * @param {string} action The action being processed
+ */
+export const handleProcessJITM = ( state, dispatch, action ) => {
+	if ( ! config.isEnabled( 'jitms' ) ) {
+		return;
+	}
+
+	if ( ! process.hasInitializedSites || ! process.hasInitializedSection ) {
+		dispatch( {
+			type: JITM_SET,
+			jitms: [],
+		} );
+		return;
+	}
+
+	const currentSite = process.lastSite;
+
+	if ( ! isJetpackSite( state, currentSite ) ) {
+		dispatch( {
+			type: JITM_SET,
+			jitms: [],
+		} );
+		return;
+	}
+
+	dispatch( http( {
+		apiNamespace: 'wpcom',
+		method: 'GET',
+		path: `/v2/sites/${ currentSite }/jitm/calypso:${ process.lastSection }:admin_notices`,
+		query: {
+			external_user_id: getCurrentUserId( state ),
+		}
+	}, action ) );
+};
+
+/**
+ * Called when a route change might have occured
+ * @param {function} getState A function to retrieve the current state
+ * @param {string} action The action being processed
+ * @param {function} dispatch A function to dispatch an action
+ */
+export const handleRouteChange = ( { getState, dispatch }, action ) => {
+	if ( ! config.isEnabled( 'jitms' ) ) {
+		return;
+	}
+
+	// wait for loading to complete before trying to show a jitm since we cannot capture enough state until
+	// loading completes
+	switch ( action.isLoading ) {
+		case false:
+			process.hasInitializedSection = true;
+			return;
+		case true:
+			process.hasInitializedSection = false;
+			return;
+	}
+
+	process.lastSection = action.section.name;
+
+	handleProcessJITM( getState(), dispatch, action );
+};
+
+/**
+ * Called when a site is selected
+ * @param {function} getState Function to get the current state
+ * @param {string} action The action being processed
+ * @param {function} dispatch The dispatch function
+ */
+export const handleSiteSelection = ( { getState, dispatch }, action ) => {
+	if ( ! config.isEnabled( 'jitms' ) ) {
+		return;
+	}
+
+	process.hasInitializedSites = ! ! action.siteId;
```

Comment:
do we want to clear this if we unselect a site? I'm guessing it's possible to see this action fire without a corresponding `siteId` (or without a truthy one)

Reply:
It will fire without a truthy `siteId` during initial load, among other times (like unselecting a site). When it does, we need to clear out the message (so that the message's context makes sense).